File: C++/Permutations_II.cpp
class Solution {
public:
    vector<vector<int> > ret;
    vector<vector<int> > permuteUnique(vector<int> &num) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        ret.clear();
        vector<int> cand;
        sort(num.begin(), num.end());
        solve(cand, num);
        return ret;
    }
    void solve(vector<int> &cand, vector<int> &remain){
        if (remain.size() == 0){
            ret.push_back(cand);
        }
        for (int n = 0; n < remain.size(); n ++){
            if (n == 0 || (n > 0 && remain[n] != remain[n-1])){
                vector<int> newCand(cand);
                vector<int> newRemain(remain);
                newCand.push_back(remain[n]);
                newRemain.erase(newRemain.begin()+n);
                solve(newCand, newRemain);
            }
        }
    }
};

File: Python/Permutations_II.py
class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
    def permuteUnique(self, num):
        num.sort()
        return self.solve(num)
        
    def solve(self, num):
        ret = []
        for i in range(len(num)):
            if i > 0 and num[i] == num[i-1]: continue
            remain = num[0:i] + num[i+1:len(num)]
            subRet = self.solve(remain)
            if len(subRet) == 0:
                ret.append([num[i]])
                continue
            for p in subRet:
                ret.append([num[i]]+p)
        return ret
